feat: native module for idle detection on wayland

T1mbits opened the pull request:
Adds a native node module for native idle detection on Wayland linux. To be used by the [NativeIdle plugin](https://github.com/Vendicated/Vencord/pull/3767).

Fixes #396

T1mbits committed:
feat: partial native idle support

Client should now in theory idle when system is suspended on all
platforms, after 10 minutes on windows, mac, and xorg (I think?), and
when the lockscreen is triggered on windows and mac. Still need to work
on implementing wayland idle and lockscreen detection, most likely in a
native module.

Only tested with NixOS with hyprland, as expected getSystemIdleTime
doesn't work and neither does lock-screen or unlock-screen events, but
suspend and resume events work fine.

T1mbits committed:
feat: native wayland idle support

Uses vesktop-wayland-protocols package to read the ext-idle-notify-v1
wayland protocol to detect idle timeouts natively.

T1mbits committed:
chore: update & mark vwp package as optional

T1mbits committed:
style: use spaces instead of tabs

T1mbits committed:
feat: CustomIdle plugin support for wayland

CustomIdle should in theory work completely for windows and macos out of
the box already (I can't really test it though)

T1mbits committed:
fix: CustomIdle disable auto-idle on wayland

Setting the CustomIdle timeout slider to 0 should now prevent you from
ever going idle automatically.

Vendicated commented:
Well it'd be a bit difficult to move it to Vencord while also having the native module, so I'm not sure about that. Don't worry about that part for now

T1mbits committed:
feat: wayland-protocols first party package

Moved vesktop-wayland-protocols source code into /packages mimicking
libvesktop for binary naming and locations, and removed
vesktop-wayland-protocols as a dependency

T1mbits commented:
Moved the package into `/packges`, seems to work fine? Again, should probably test on other platforms and compositors. If you don't trust the prebuilt binaries you can feel free to rebuild them yourself.

T1mbits committed:
Merge remote-tracking branch 'upstream/main' into native-idle

T1mbits commented:
it’s an unfixable problem for wayland, there isn’t a widely used protocol that allows clients to know if the system is locked or not so lockscreen idling is exclusive to windows and macos :/

nixRidge commented:
Got it, makes sense to me. So when the lockscreen is active (at least on Plasma as per the example), it just won't ever detect the user as idle?

In any case it's not a big issue.
I'm working around it by using the CustomIdle plugin and making it set my status to idle just before the screen locks (4,8 minutes), and in that case everything is fine. It even sets my status back to online when I move my mouse/click on the lockscreen.

nixRidge commented:
I'll do that tomorrow. Thanks for letting me know!

T1mbits committed:
feat(waylandIdle): move functionality into a Vencord plugin

Vesktop still needs to provide the native node module for Wayland

T1mbits committed:
fix(waylandIdle): set idle timer to use minutes instead of seconds

MatthewCash commented:
> Moved the package into `/packges`, seems to work fine? Again, should probably test on other platforms and compositors. If you don't trust the prebuilt binaries you can feel free to rebuild them yourself.

I don't see how this is better, its still just a prebuilt binary manually committed into the tree? I think the idea was that building libvesktop should also build the node addon and that no binaries are committed or downloaded.

MatthewCash commented:
> The maintainers are also free to completely replace the binaries with their own compiled versions if they want. I wrote it in rust because, as I said, I couldn't write in C++ to extend libvesktop. If you want you can rewrite the module into libvesktop yourself to circumvent whatever issue you see with the way it currently is, but I unfortunately can't as of right now.

I might give it a shot if I get some time today :)

MatthewCash commented:
I added a prototype C++ port of your code to libvesktop in be04d6bef0ceb732d6c918cfbaf6f9138b818c96, so far completely untested or integrated, it builds though. It's missing the destructor for `IdleNotifier`, I'm not sure what the best approach for that is yet.

MatthewCash committed:
Port wayland-protocols Rust module to libvesktop in C++

T1mbits committed:
Merge branch 'main' into native-idle

T1mbits commented:
@MatthewCash not sure if you're still interested in this, but I've started working on this again and I've pulled your commit into my fork, hopefully that's okay with you. I haven't fully tested/recompiled your implementation yet as I've been busy reconstructing the powerMonitor events and basically remaking the plugin from scratch but I thought I should probably keep you updated on this.

T1mbits committed:
feat(idle): add powerMonitor callbacks to VesktopNative

Rename waylandIdle to nativeIdle since it's not just wayland stuff
anymore. This is needed so that callbacks can be made by the plugin
on powerMonitor events since I can't seem to get it to work with just
Vencord native plugins.

MatthewCash commented:
I don't really have time to work on this anymore so I'm just hoping you can get this fixed :)

T1mbits committed:
fix(libvesktop): use gitlab for ext-idle-notify-v1 xml source

The previous link was a 404 on a mirror?

T1mbits committed:
fix(libvesktop): fully register idle notifier objects with compositor

Also no deconstructor still so Vesktop errors when you close it if the
idle module was loaded lol. I'll add that alongside a proper on resume
callback (although that looks partially implemented already :O) soon

T1mbits commented:
Today I learned that I am not built for c++. I am a spoiled child and I need my cargo and my compiler and my docs.rs. 4 hours of headaches and I still can't get a destructor working for IdleNotifier. I'm probably just going to put back my rust implementation with proper integration into the buildsystem (with docker and build scripts for prebuilts and the likes), unless somebody else who is actually capable and has free time can patch things up.

I'll wait till the end of the week before doing anything but if I do I'll probably try to implement the protocols @eth-p mentioned earlier.
